### ISSUE TITLE
Discrepancies between  CLBlast::gemmStridedBatched and CLBlast::gemm

### ISSUE THREAD

**Issue by hshatti** [CNugteren/CLBlast#578]
Hi!, 
I have noticed  that calling a strided batch of general matrix multiplications with a single call using the function <b>CLBlastSgemmStridedBatched</b> instead of looping through <b>CLBlastSgemm</b> is significantly faster, in some cases X2 which is great, however!, by time I have discovered a significant discrepancies between the two function's results, <b>CLBlastSgemm</b> produces a correct results while <b>CLBlastSgemmStridedBatched</b> produces a different (incorrect) results in C matrix, I'm not sure if these inaccuracies exists in other batch functions but it seems that ( and perhaps I'm mistaking) the scaler <b>BETA</b> argument doesn't scale the matrix C before performing the gemm OP or maybe it's something else? here is an example code to reproduce the issue, my original code using CLblast was written in Delphi when I discovered this (headers ported to object pascal) however!,  you can check the problem as in the following C++ code snippet that I have rewritten reproducing the same issue.

Edit : just made some experiments and it seems that strided function fails only when setting a STRIDE for C matrix to zero, while in theory as in cuBLAS or Intel's MKL it they accumulate the results in one matrix, I don't know if this is a normal behaviour  

```c++

// include CL and clblast  headers and stuff 1st...

#include <stdio.h>
#include <vector>
#include <cmath>
// etc...

//...
using namespace std; // and that's OK

typedef vector<float> Matrix;

void rand_matrix(Matrix& m){
  for(int i=0; i<m.size(); i++)
    m[i] = static_cast<float>((rand()%2048))/1024 -1. ;  // between  -1.0 and +1.0
}

float sum_square_difference(const Matrix a, const Matrix b){
  float result = 0;
  for (int i=0 ; i<a.size(); i++)
    result += pow(a[i]-b[i], 2);
  return result;
}


struct CLStuff{

public:
  CLStuff(){
    cl_int err;
    err = clGetPlatformIDs(1, &platform, nullptr);
    err = clGetDeviceIDs(platform, CL_DEVICE_TYPE_GPU, 1, &device, nullptr);
    context = clCreateContext(nullptr, 1, &device, nullptr, nullptr, &err);
    queue = clCreateCommandQueue(context, device, 0, &err);
  }

  cl_platform_id platform;
  cl_device_id device;
  cl_context context;
  cl_command_queue queue;
  cl_mem clAlloc(const size_t bytes){
    cl_int err;
    return clCreateBuffer(context, CL_MEM_READ_WRITE, bytes, nullptr, &err);
  }

  void clPull(const size_t bytes, const cl_mem src, void* host){
    clEnqueueReadBuffer(queue, src, CL_TRUE, 0, bytes, host, 0, nullptr, nullptr);
  }
  void clPush(const size_t bytes, const void* src, cl_mem dev){
    clEnqueueWriteBuffer(queue, dev, CL_TRUE, 0, bytes, src, 0, nullptr, nullptr);
  }
};

#define MATRIX_BATCH 32
#define WIDTH 100
#define HEIGHT 100
#define STRIDE WIDTH*HEIGHT 
int main(int argc, char** argv){

  srand(time(0));  //  initialize random seed
  CLStuff ocl = CLStuff(); // maybe instantiation isn't required!, but just to be sure ;

  Matrix A    = Matrix(MATRIX_BATCH * WIDTH * HEIGHT);
  rand_matrix(A); // fill random floats between -1 and 1
  cl_mem devA = ocl.clAlloc(A.size()*sizeof(float));
  ocl.clPush(A.size()*sizeof(float), A.data(), devA);
 
  Matrix B    = Matrix(MATRIX_BATCH * WIDTH * HEIGHT);
  rand_matrix(B); // fill random floats between -1 and 1
  cl_mem devB = ocl.clAlloc(B.size()*sizeof(float));
  ocl.clPush(B.size()*sizeof(float), B.data(), devB);

  Matrix C1    = Matrix(WIDTH * HEIGHT);
  rand_matrix(C1);  // fill random floats between -1 and 1
  cl_mem devC1 = ocl.clAlloc(C1.size()*sizeof(float));
  ocl.clPush(C1.size()*sizeof(float), C1.data(), devC1);

  Matrix C2(C1);
  cl_mem devC2 = ocl.clAlloc(C2.size()*sizeof(float));
  ocl.clPush(C2.size()*sizeof(float), C2.data(), devC2);


 // calculate  gemm using strided batch ( inaccurate result in devC1)
  CLBlastSgemmStridedBatched(
    CLBlastLayoutRowMajor, CLBlastTransposeNo, CLBlastTransposeYes,  // second matricies (B) are transposed
    HEIGHT, WIDTH, WIDTH, 
    1., devA, 0, WIDTH, STRIDE, 
    devB, 0, WIDTH, STRIDE,
    1., devC1, 0, WIDTH, 0, // stride is 0  which means accumulate multiplication results to one matrix 
    MATRIX_BATCH, &ocl.queue, nullptr
  );

 // same calculation but using  a for loop ( correct result in devC2 )
  
  for (int i=0; i<MATRIX_BATCH; i++)
    CLBlastSgemm(
      CLBlastLayoutRowMajor, CLBlastTransposeNo, CLBlastTransposeYes,  // second matricies (B) are transposed
      HEIGHT, WIDTH, WIDTH, 
      1., devA, i*STRIDE, WIDTH,  
      devB, i*STRIDE, WIDTH,
      1., devC2, 0, WIDTH, // stride is 0  which means accumulate multiplication results to one matrix 
      &ocl.queue, nullptr
    );

 
  ocl.clPull(C1.size()*sizeof(float), devC1, C1.data());
  ocl.clPull(C2.size()*sizeof(float), devC2, C2.data());

  printf("sumSquareDiff(A,B) : %f\n", sum_square_difference(C1, C2));
  // will print a non zero number indicating discrepancies 


  return 0;
}
```
Regards
H

**Comment by CNugteren** [comment]
Thanks for reporting this issue and sharing example code.

First thing to do would be to run the official CLBlast test suite on your device, i.e. `./clblast_test_xgemmstridedbatched --full_test`. You can even add `--verbose` such that it prints what it is testing, such that you can see what kind of combinations and values of parameters are tested and which ones aren't.

> Edit : just made some experiments and it seems that strided function fails only when setting a STRIDE for C matrix to zero, while in theory as in cuBLAS or Intel's MKL it they accumulate the results in one matrix, I don't know if this is a normal behaviour

If that is indeed the only error that you could find, then it is likely that I didn't design this function to work will a stride for C of zero, i.e. `ldc=0`. I don't think that is tested either. But, you are right that this is a bug: either it should return an error code saying that this isn't supported, or it should actually work.

**Comment by EJain-Dev** [comment]
Created pull request #583 to report that 'ldc=0' is not supported with a new error code.

**Comment by hshatti** [comment]
Hi,
just to clarify, what I meant by stride of C is the batch stride <c_stride> not the matrix width stride <ldc>, in theory its a logical fix to throw an exception if <ldc> was set to Zero but in the case of setting the batch stride of C <c_stride> to Zero, instead of raising an exception, the suggested solution could be for example like in the other BLAS libraries, is to perform an element wise sum of production results into one matrix which is C, this special case is really handy and perfectly fit's the purpose for example during the backpropagation of convolutional neural networks where it would be ideal to calculate the weights gradient matrices of a batch with one call of this function and let the GPU do the rest instead of looping through the batch, however the function performance it self is spectacularly fast! and this special case a minor suggestion, I could implement a fix for this and submit a pull request with a test for it sometime if you think it would be a good idea :)

**Comment by EJain-Dev** [comment]
If you want to make a cross-platform OpenCL AI & ML library, ArrayFire implements a lot of things already and supports various backends (CUDA, OpenCL with CLBlast, SYCL, oneAPI, ROCM, CPU, etc.).

**Comment by CNugteren** [comment]
Thanks for clarifying. Apologies for the misunderstanding. Yes I would be happy to review and accept a PR that adds this feature!
